src/java/org/apache/cassandra/streaming/async/StreamCompressionSerializer.java

```diff
@@ -0,0 +1,140 @@
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.memory.BufferPool;
+
+/**
+ * A serialiazer for stream compressed files (see package-level documentation). Much like a typical compressed
+ * output stream, this class operates on buffers or chunks of the data at a a time. The format for each compressed
+ * chunk is as follows:
+ *
+ * - int - compressed payload length
+ * - int - uncompressed payload length
+ * - bytes - compressed payload
+ */
+public class StreamCompressionSerializer
+{
+    public static final StreamCompressionSerializer serializer = new StreamCompressionSerializer();
+
+    private StreamCompressionSerializer()
+    {   }
+
+    /**
+     * Length of heaer data, which includes compressed length, uncompressed length.
+     */
+    private static final int HEADER_LENGTH = 8;
+
+    /**
+     *
+     * @return A buffer with decompressed data. The returned buffer is possibly taken from the {@link BufferPool}, and
+     * thus you need call {@link BufferPool#put(ByteBuffer)} when it has been consumed to ensure the buffer
+     * is returned to the pool.
+     */
+    public ByteBuffer serialize(LZ4Compressor compressor, ByteBuffer in, int version)
+    {
+        final int uncompressedLength = in.remaining();
+
+        int maxLength = compressor.maxCompressedLength(uncompressedLength);
+        ByteBuffer compressed = BufferPool.get(maxLength);
+        try
+        {
+            compressor.compress(in, compressed);
+            int compressedLength = compressed.position();
+            compressed.limit(compressedLength).position(0);
+
+            // not sure if it's better to alloc a CompositeByteBuf and add the two buffers to it,
+            // or create another ByteBuf and copy the compressed data to it (after the headers)
+
+            ByteBuffer out = BufferPool.get(HEADER_LENGTH + compressedLength);
+            out.putInt(compressedLength);
+            out.putInt(uncompressedLength);
+            out.put(compressed);
+            out.flip();
+            return out;
+        }
+        finally
+        {
+            BufferPool.put(compressed);
+        }
+    }
+
+    /**
+     *
+     * @return A buffer with decompressed data. The returned buffer is possibly taken from the {@link BufferPool}, and
+     * thus you need call {@link BufferPool#put(ByteBuffer)} when it has been consumed to ensure the buffer
+     * is returned to the pool.
+     */
+    public ByteBuffer deserialize(LZ4FastDecompressor decompressor, DataInputPlus in, int version) throws IOException
+    {
+        final int compressedLength = in.readInt();
+        final int uncompressedLength = in.readInt();
+
+        if (in instanceof ReadableByteChannel)
+        {
+            ByteBuffer compressed = BufferPool.get(compressedLength);
+            ByteBuffer uncompressed = null;
+            try
+            {
+                int readLength = ((ReadableByteChannel) in).read(compressed);
```

Comment:
huh, maybe i didn't click "Comment" before, and I lost my comments :(

The reason why I'm copying here is that there's no guarantee the bytes you want to read (in the `in` paramaeter, basically a `RebufferingByteBufDataInputPlus`) are in one contiguous buffer. So you'll have to copy them into one buffer at some point if they are different buffers.

Thus, I thought implementing the `ReadableByteChannel` interface might be the least crappy solution for copying (on `RebufferingByteBufDataInputPlus`)